### PR TITLE
Pixel template local reconstruction for Phase 1

### PR DESCRIPTION
**Pull request by pmaksim1**
This PR is essentially the same as #17502 , except that diff would say that everything has changed since almost all the code was consistently reformatted.   We also replaced one ;; with a ;.

We also commented out the use of 2D templates in template splitter, since it was not being used anywhere.  (It might one day, but it is not needed now.)

The code has been tested on
- Phase 1 MC
- Run 1 data.  The code is backwards compatible since we get identical results.
(More details at the Pixel Offline meeting at https://indico.cern.ch/event/607516/)

To run on 0T or 3.8T data, the code needs to load the new templates.

### CODE REVIEW

**Comment by slava77** [comment]
@pmaksim1 
global re-indentation will make code history navigation (via "blame") pretty complicated.
Was it agreed with other RecoLocalTracker/SiPixelRecHits developers?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by boudoul** [comment]
Hi @pmaksim1 , I would second @slava77 comment , could you please try not to change the indentation when it's not needed ? at it is now, looking at the diff 
https://github.com/cms-sw/cmssw/pull/18038/files
is making the undertanding of your changes very complicated ....

**Comment by slava77** [comment]
On 3/22/17 10:10 AM, boudoul wrote:
> Hi @pmaksim1 <https://github.com/pmaksim1> , I would second @slava77
> <https://github.com/slava77> comment , could you please try not to
> change the indentation when it's not needed ? at it is now, looking at
> the diff
> https://github.com/cms-sw/cmssw/pull/18038/files
> is making the undertanding of your changes very complicated ....

The PR review is "safe" by adding ?w=1 to the URL.
There are only +349 −101 lines changed in 19 files
(not +8,136 −7,888 without this empty/white-space suppression)

**Comment by pmaksim1** [comment]
Yes.  This was discussed within the Pixel reco group.

That is why we re-did all the tests even though there was no actual change
in the code.

                 Petar

**Comment by pmaksim1** [comment]
Well, initially, I was reluctant to make any purely cosmetic changes.

But the fact is that the lack of proper indentation on Linux was making the
code hard to read, and thus to support. (Note that the code was indented
just fine on a Mac, which is where most of it has been and is being
developed.)  In the long run, this had to be indented *at some point*.

So Morris and I decided that now is the time we bite the bullet.

This is like rewriting the code to make the organization and logical flow
more legible, and thus more supportable.  It can be postponed, but at some
point it should be done.  So sooner is better than later.

                   Petar

**Comment by boudoul** [comment]
ok thanks @pmaksim1 ! and also thanks for the trick of ?w=1   added to my list of tricks :)

**Comment by slava77** [comment]
@pmaksim1 

based on the PR description, it sounds like I should expect no changes in run1 plots. Instead, there are (small) changes in data and MC in run1 and run2.
Can this all be blamed to some floating point math reshuffling (refactoring of the same formulas or changes in precision) or were there any bugfixes as well?

**Comment by pmaksim1** [comment]
Hi Slava,

there were a couple of rather minor bug fixes/improvements that we picked
up.  One of them, with the comment "Fix centering problem for small
signals"  could create differences in a large distribution.  The point is
that the results won't be bitwise identical with the old ones.  (I don't
actually know what's the reference sample is.)

We saw this and didn't worry much about it since the effect is small and
the current code still runs OK on Run 1 data.

Hope this helps!

            Petar

File: RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco.h [hunk 1]
```diff
@@ -50,49 +53,52 @@
...
+#include "SiPixelTemplateDefs.h"
 #include "SiPixelTemplate.h"
 #endif
 
 #define N2D 500
 
 #include <vector>
 
+#ifndef SiPixelTemplateClusMatrix
+#define SiPixelTemplateClusMatrix 1
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco.h]: why is this needed?
I did not find any use cases of ``` SiPixelTemplateClusMatrix ```

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc [hunk 2]
```diff
@@ -111,2143 +114,2222 @@ using namespace edm;
...
+         
+         if(in_file.fail()) {LOGERROR("SiPixelTemplate") << "Error reading file 0B, no template load"
+            << ENDL; return false;}
+      } else {
+         theCurrentTemp.head.ss50 = theCurrentTemp.head.s50;
+         theCurrentTemp.head.lorybias = theCurrentTemp.head.lorywidth/2.f;
+         theCurrentTemp.head.lorxbias = theCurrentTemp.head.lorxwidth/2.f;
+         theCurrentTemp.head.fbin[0] = 1.5f;
+         theCurrentTemp.head.fbin[1] = 1.00f;
+         theCurrentTemp.head.fbin[2] = 0.85f;
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: these magic numbers (used in multiple places) were constants before 
```    const float fbin0=1.50f, fbin1=1.00f, fbin2=0.85f; ```
these should stay using constants, unless there is some rationale not to.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc [hunk 3]
```diff
@@ -111,2143 +114,2222 @@ using namespace edm;
...
-	    
-	    for (j=0; j<4; ++j) {
-	      
-	      db >> theCurrentTemp.enty[i].xavggen[j] >> theCurrentTemp.enty[i].xrmsgen[j] >> theCurrentTemp.enty[i].xgx0gen[j] >> theCurrentTemp.enty[i].xgsiggen[j];
-	      
-	      if(db.fail()) {LOGERROR("SiPixelTemplate") << "Error reading file 14b, no template load, run # " << theCurrentTemp.enty[i].runnum << ENDL; return false;}
-	    } 
-	    
-	    
-	    db >> theCurrentTemp.enty[i].chi2yavgone >> theCurrentTemp.enty[i].chi2yminone >> theCurrentTemp.enty[i].chi2xavgone >> theCurrentTemp.enty[i].c
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: why not initialize the line above?

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc [hunk 4]
```diff
@@ -2263,237 +2345,309 @@ void SiPixelTemplate::xtemp3d(int i, int j, std::vector<float>& xtemplate)
...
+int SiPixelTemplate::qbin(int id, float cotalpha, float cotbeta, float locBz, float qclus, float& pixmx,
+                          float& sigmay, float& deltay, float& sigmax, float& deltax,float& sy1, float& dy1, float& sy2,
+                          float& dy2, float& sx1, float& dx1, float& sx2, float& dx2) // float& lorywidth, float& lorxwidth)
+{
+   // Interpolate for a new set of track angles
+   
+   // Local variables
+   float locBx; //  lorywidth, lorxwidth;
+   // Local variables
+   locBx = 1.f;
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: why not initialize together with declaration?

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc [hunk 5]
```diff
@@ -72,7 +75,7 @@ static const int theVerboseLevel = 2;
 #include "FWCore/Utilities/interface/Exception.h"
 #else
 #include "SiPixelTemplateReco.h"
-#include "VVIObj.h"
+#include "VVIObjF.h"
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc]: Remind me please what is the difference between the two.
Thank you.

**Comment by pmaksim1** [comment]
Hi Slava,

most of these lines are in Morris's code -- I'll sit down with him later
today, discuss, and get back to you.

                Petar

**Comment by slava77** [comment]
Hi Petar,

I checked comparisons of this PR effect with respect to 910pre1 and things look OK (only small changes). Timing is also roughly unchanged.

So, the only remaining items are from the code review.

Slava

**Comment by pmaksim1** [comment]
Hi Slava,

Morris escaped to CERN, so I'll hunt him down on Wednesday and hopefully we
can wrap this up.

              Petar

File: RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco.h (hunk 1 above)
Comment (pmaksim1) [inline RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco.h]: Sorry for the delay.  Morris and I looked at all of these, so here are responses one at the time.

This one is used to block duplicate inclusion the definition of ClusMatrix object.  (Why it wasn't called SiPixelTemplateClusMatrix, I don't know.)  This was added by Vincenzo I. some time ago.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc (hunk 2 above)
Comment (pmaksim1) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: So, the magic numbers show up only twice, as the default values for some of the template parameters.  Usually a const would be used, if we expected to *change* these magic values.  However, the values themselves are needed for *old* templates -- the new template objects have numbers read from the template DB payload.  The old templates are frozen in time and these numbers will not change.

So we can add a const for cosmetic purposes, but we don't think it's needed here.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc (hunk 3 above)
Comment (pmaksim1) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: Both Morris and I were somewhat confused by this.  Would you mind giving us a bit more info to work with?

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc (hunk 4 above)
Comment (pmaksim1) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: We agree.  (On an analysis review twiki this would be color-coded as green :)

If we are making other changes, we'll do this.  However, if this is the only change, we maybe should defer this particular (and trivial) fix until the next PR.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc (hunk 5 above)
Comment (pmaksim1) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc]: VVIObj is Morris's port of the original CERNLIB Fortran code.  All the variables are double.

Some time ago Vincenzo I. rewrote it to use floats instead of doubles.  It's slightly faster.  That's VVIObjF.

However, note that this change is in the *standalone* branch, which is used for standalone testing of all the templates and other pixel local reco.  (Standalone = sans CMSSW.)  So this particular change is actually a bugfix which ensures that the standalone version used for development and testing is actually the same as used in CMSSW.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc (hunk 2 above)
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: just to clarify on my comment, your change here, once you get past all the reindentation, is 
```diff
-	   theCurrentTemp.head.fbin[0] = fbin0;
-	   theCurrentTemp.head.fbin[1] = fbin1;
-	   theCurrentTemp.head.fbin[2] = fbin2;
+         theCurrentTemp.head.fbin[0] = 1.5f;
+         theCurrentTemp.head.fbin[1] = 1.00f;
+         theCurrentTemp.head.fbin[2] = 0.85f;
```
So, I'm just asking to go back to what was here before since it's a more appropriate style for magic numbers.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc (hunk 3 above)
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: the comment was made for a different line

Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: this comment https://github.com/cms-sw/cmssw/pull/18038#pullrequestreview-30247938 about initialization on the same line was meant for this line (new version line 1499)

File: RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco.h (hunk 1 above)
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/interface/SiPixelTemplateReco.h]: There is an include guard for this file already and ```SiPixelTemplateClusMatrix``` is not appearing anywhere else.
So, this ifndef has no effect in current CMSSW and looks more like a source of confusion

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/18038	d0b16ee
- code changes are in line with the description and comments made in the previous PR
- jenkins tests pass and show small changes in tracking quantities in workflow where the pixel templates are used (run1/2; not in phase-1 2017)
- local tests with larger samples confirm that the changes are small

The agreement with @pmaksim1 is that there will be a technical follow up PR to address (mostly stylistic) code review issues. There is a github issue for this #18227

**Comment by davidlange6** [comment]
+1
